File: frontend/src/utils/marketHours.ts
export interface ExchangeInfo {
  timezone: string;
  // Regular session
  openHour: number;
  openMin: number;
  closeHour: number;
  closeMin: number;
  // Extended sessions (optional)
  preMarketOpenHour?: number;
  preMarketOpenMin?: number;
  afterHoursCloseHour?: number;
  afterHoursCloseMin?: number;
  name: string;
  region: 'Americas' | 'Europe' | 'Asia-Pacific';
}

const US_ET = 'America/New_York';

export const EXCHANGE_HOURS: Record<string, ExchangeInfo> = {
  // Americas — US (IBKR pre-market 04:00–09:30, after-hours 16:00–20:00 ET)
  NYSE:   { timezone: US_ET, openHour: 9, openMin: 30, closeHour: 16, closeMin: 0, preMarketOpenHour: 4, preMarketOpenMin: 0, afterHoursCloseHour: 20, afterHoursCloseMin: 0, name: 'New York Stock Exchange',  region: 'Americas' },
  NASDAQ: { timezone: US_ET, openHour: 9, openMin: 30, closeHour: 16, closeMin: 0, preMarketOpenHour: 4, preMarketOpenMin: 0, afterHoursCloseHour: 20, afterHoursCloseMin: 0, name: 'NASDAQ',                   region: 'Americas' },
  ARCA:   { timezone: US_ET, openHour: 9, openMin: 30, closeHour: 16, closeMin: 0, preMarketOpenHour: 4, preMarketOpenMin: 0, afterHoursCloseHour: 20, afterHoursCloseMin: 0, name: 'NYSE Arca',                region: 'Americas' },
  AMEX:   { timezone: US_ET, openHour: 9, openMin: 30, closeHour: 16, closeMin: 0, preMarketOpenHour: 4, preMarketOpenMin: 0, afterHoursCloseHour: 20, afterHoursCloseMin: 0, name: 'NYSE American',            region: 'Americas' },
  BATS:   { timezone: US_ET, openHour: 9, openMin: 30, closeHour: 16, closeMin: 0, preMarketOpenHour: 4, preMarketOpenMin: 0, afterHoursCloseHour: 20, afterHoursCloseMin: 0, name: 'Cboe BZX (BATS)',          region: 'Americas' },
  IEX:    { timezone: US_ET, openHour: 9, openMin: 30, closeHour: 16, closeMin: 0, preMarketOpenHour: 4, preMarketOpenMin: 0, afterHoursCloseHour: 20, afterHoursCloseMin: 0, name: 'Investors Exchange',       region: 'Americas' },
  SMART:  { timezone: US_ET, openHour: 9, openMin: 30, closeHour: 16, closeMin: 0, preMarketOpenHour: 4, preMarketOpenMin: 0, afterHoursCloseHour: 20, afterHoursCloseMin: 0, name: 'IBKR Smart Routing (US)',  region: 'Americas' },
  CBOE:   { timezone: US_ET, openHour: 9, openMin: 30, closeHour: 16, closeMin: 0, preMarketOpenHour: 4, preMarketOpenMin: 0, afterHoursCloseHour: 20, afterHoursCloseMin: 0, name: 'Cboe Options Exchange',    region: 'Americas' },
  // Americas — Canada (TSX pre-market 07:00–09:30, after-hours 16:00–17:00 ET)
  TSX:    { timezone: 'America/Toronto', openHour: 9, openMin: 30, closeHour: 16, closeMin: 0, preMarketOpenHour: 7, preMarketOpenMin: 0, afterHoursCloseHour: 17, afterHoursCloseMin: 0, name: 'Toronto Stock Exchange', region: 'Americas' },
  TSXV:   { timezone: 'America/Toronto', openHour: 9, openMin: 30, closeHour: 16, closeMin: 0, preMarketOpenHour: 7, preMarketOpenMin: 0, afterHoursCloseHour: 17, afterHoursCloseMin: 0, name: 'TSX Venture Exchange',   region: 'Americas' },
  // Europe — London (pre-auction 07:00–08:00, closing auction ends ~16:35)
  LSE:    { timezone: 'Europe/London',    openHour: 8, openMin: 0, closeHour: 16, closeMin: 30, preMarketOpenHour: 7, preMarketOpenMin: 0, afterHoursCloseHour: 17, afterHoursCloseMin: 15, name: 'London Stock Exchange',    region: 'Europe' },
  IOB:    { timezone: 'Europe/London',    openHour: 8, openMin: 0, closeHour: 16, closeMin: 30, preMarketOpenHour: 7, preMarketOpenMin: 0, afterHoursCloseHour: 17, afterHoursCloseMin: 15, name: 'LSE Intl Order Book',      region: 'Europe' },
  // Europe — Xetra/Frankfurt (pre-trading 08:00–09:00, post-trading 17:30–20:00)
  XETRA:  { timezone: 'Europe/Berlin',    openHour: 9, openMin: 0, closeHour: 17, closeMin: 30, preMarketOpenHour: 8, preMarketOpenMin: 0, afterHoursCloseHour: 20, afterHoursCloseMin: 0, name: 'Xetra (Frankfurt)',        region: 'Europe' },
  FWB:    { timezone: 'Europe/Berlin',    openHour: 9, openMin: 0, closeHour: 17, closeMin: 30, preMarketOpenHour: 8, preMarketOpenMin: 0, afterHoursCloseHour: 20, afterHoursCloseMin: 0, name: 'Frankfurt Stock Exchange', region: 'Europe' },
  // Europe — Euronext (pre-opening 07:15–09:00, TAL session 17:30–17:40)
  SBF:    { timezone: 'Europe/Paris',     openHour: 9, openMin: 0, closeHour: 17, closeMin: 30, preMarketOpenHour: 7, preMarketOpenMin: 15, name: 'Euronext Paris',           region: 'Europe' },
  AEB:    { timezone: 'Europe/Amsterdam', openHour: 9, openMin: 0, closeHour: 17, closeMin: 30, preMarketOpenHour: 7, preMarketOpenMin: 15, name: 'Euronext Amsterdam',       region: 'Europe' },
  // Asia-Pacific — no standardised extended sessions
  ASX:    { timezone: 'Australia/Sydney', openHour: 10, openMin: 0, closeHour: 16, closeMin: 0,  name: 'Australian Securities Exch.',  region: 'Asia-Pacific' },
  TSEJ:   { timezone: 'Asia/Tokyo',       openHour: 9,  openMin: 0, closeHour: 15, closeMin: 30, name: 'Tokyo Stock Exchange',         region: 'Asia-Pacific' },
  OSE:    { timezone: 'Asia/Tokyo',       openHour: 9,  openMin: 0, closeHour: 15, closeMin: 30, name: 'Osaka Exchange',               region: 'Asia-Pacific' },
  SEHK:   { timezone: 'Asia/Hong_Kong',  openHour: 9,  openMin: 30, closeHour: 16, closeMin: 0, name: 'Hong Kong Exchanges',          region: 'Asia-Pacific' },
  SGX:    { timezone: 'Asia/Singapore',   openHour: 9,  openMin: 0, closeHour: 17, closeMin: 0,  name: 'Singapore Exchange',           region: 'Asia-Pacific' },
  NSE:    { timezone: 'Asia/Kolkata',     openHour: 9,  openMin: 15, closeHour: 15, closeMin: 30, name: 'National Stock Exch. India',  region: 'Asia-Pacific' },
  BSE:    { timezone: 'Asia/Kolkata',     openHour: 9,  openMin: 15, closeHour: 15, closeMin: 30, name: 'Bombay Stock Exchange',        region: 'Asia-Pacific' },
};

const DEFAULT = EXCHANGE_HOURS['NYSE'];

export type SessionType = 'pre-market' | 'open' | 'after-hours' | 'closed';

export interface MarketStatus {
  session: SessionType;
  isOpen: boolean;   // true only during the regular session
  countdown: string;
}

function _localNow(tz: string): Date {
  return new Date(new Date().toLocaleString('en-US', { timeZone: tz }));
}

function _fmtMins(totalMins: number): string {
  if (totalMins <= 0) return '< 1m';
  const h = Math.floor(totalMins / 60);
  const m = totalMins % 60;
  if (h === 0) return `${m}m`;
  if (m === 0) return `${h}h`;
  return `${h}h ${m}m`;
}

function _minsUntilNextWeekdaySession(currentMins: number, day: number, targetMins: number): number {
  // Called when we're past the target time today (or it's a weekend).
  let nextDay = (day + 1) % 7;
  let daysAhead = 1;
  while (nextDay === 0 || nextDay === 6) { nextDay = (nextDay + 1) % 7; daysAhead++; }
  return (24 * 60 - currentMins) + (daysAhead - 1) * 24 * 60 + targetMins;
}

export function getMarketStatus(exchange: string): MarketStatus {
  const spec = EXCHANGE_HOURS[exchange.toUpperCase()] ?? DEFAULT;
  const now = _localNow(spec.timezone);
  const day = now.getDay();
  const cur = now.getHours() * 60 + now.getMinutes();
  const openMins  = spec.openHour  * 60 + spec.openMin;
  const closeMins = spec.closeHour * 60 + spec.closeMin;
  const preMins   = spec.preMarketOpenHour !== undefined ? spec.preMarketOpenHour * 60 + (spec.preMarketOpenMin ?? 0) : null;
  const ahMins    = spec.afterHoursCloseHour !== undefined ? spec.afterHoursCloseHour * 60 + (spec.afterHoursCloseMin ?? 0) : null;
  const isWeekday = day >= 1 && day <= 5;

  if (isWeekday) {
    if (cur >= openMins && cur < closeMins)
      return { session: 'open', isOpen: true, countdown: `Closes in ${_fmtMins(closeMins - cur)}` };
    if (preMins !== null && cur >= preMins && cur < openMins)
      return { session: 'pre-market', isOpen: false, countdown: `Regular opens in ${_fmtMins(openMins - cur)}` };
    if (ahMins !== null && cur >= closeMins && cur < ahMins)
      return { session: 'after-hours', isOpen: false, countdown: `Extended ends in ${_fmtMins(ahMins - cur)}` };
  }

  // Closed — count down to next session start (pre-market if defined, else regular open)
  const nextSessionMins = preMins ?? openMins;
  let minsUntil: number;
  if (isWeekday && cur < nextSessionMins) {
    minsUntil = nextSessionMins - cur;
  } else {
    minsUntil = _minsUntilNextWeekdaySession(cur, day, nextSessionMins);
  }

  const h = Math.floor(minsUntil / 60);
  const d = Math.floor(h / 24);
  const label = d > 0
    ? `Opens in ${d}d ${h % 24}h ${minsUntil % 60}m`
    : `Opens in ${_fmtMins(minsUntil)}`;

  return { session: 'closed', isOpen: false, countdown: label };
}

/** True only during the regular session — used by automated strategy guards. */
export function isMarketHours(exchange = 'NYSE'): boolean {
  const spec = EXCHANGE_HOURS[exchange.toUpperCase()] ?? DEFAULT;
  const now = _localNow(spec.timezone);
  const day = now.getDay();
  if (day === 0 || day === 6) return false;
  const mins = now.getHours() * 60 + now.getMinutes();
  return mins >= spec.openHour * 60 + spec.openMin && mins < spec.closeHour * 60 + spec.closeMin;
}

/** True during pre-market, regular, or after-hours — used by manual trade guards. */
export function isTradingSession(exchange = 'NYSE'): boolean {
  return getMarketStatus(exchange).session !== 'closed';
}

File: backend/app/utils/market_hours.py
"""Market hours utility — exchange-aware, shared by strategy runner, orders API, and bridge.

Holidays are not accounted for; this is a best-effort guard against overnight/weekend noise.
Lunch-break markets (Tokyo, Hong Kong, China) are treated as single sessions for simplicity.
"""

from datetime import datetime, time
from zoneinfo import ZoneInfo
from dataclasses import dataclass


@dataclass(frozen=True)
class _Hours:
    timezone: str
    open: time
    close: time


_MAP: dict[str, _Hours] = {
    # US equities / options
    "NYSE":   _Hours("America/New_York",  time(9, 30),  time(16, 0)),
    "NASDAQ": _Hours("America/New_York",  time(9, 30),  time(16, 0)),
    "ARCA":   _Hours("America/New_York",  time(9, 30),  time(16, 0)),
    "AMEX":   _Hours("America/New_York",  time(9, 30),  time(16, 0)),
    "BATS":   _Hours("America/New_York",  time(9, 30),  time(16, 0)),
    "IEX":    _Hours("America/New_York",  time(9, 30),  time(16, 0)),
    "SMART":  _Hours("America/New_York",  time(9, 30),  time(16, 0)),
    "CBOE":   _Hours("America/New_York",  time(9, 30),  time(16, 0)),
    # Canada
    "TSX":    _Hours("America/Toronto",   time(9, 30),  time(16, 0)),
    "TSXV":   _Hours("America/Toronto",   time(9, 30),  time(16, 0)),
    # United Kingdom
    "LSE":    _Hours("Europe/London",     time(8, 0),   time(16, 30)),
    "IOB":    _Hours("Europe/London",     time(8, 0),   time(16, 30)),
    # Germany
    "XETRA":  _Hours("Europe/Berlin",     time(9, 0),   time(17, 30)),
    "FWB":    _Hours("Europe/Berlin",     time(9, 0),   time(17, 30)),
    # France
    "SBF":    _Hours("Europe/Paris",      time(9, 0),   time(17, 30)),
    # Netherlands
    "AEB":    _Hours("Europe/Amsterdam",  time(9, 0),   time(17, 30)),
    # Australia
    "ASX":    _Hours("Australia/Sydney",  time(10, 0),  time(16, 0)),
    # Japan (simplified: no lunch break 11:30–12:30)
    "TSEJ":   _Hours("Asia/Tokyo",        time(9, 0),   time(15, 30)),
    "OSE":    _Hours("Asia/Tokyo",        time(9, 0),   time(15, 30)),
    # Hong Kong (simplified: no lunch break 12:00–13:00)
    "SEHK":   _Hours("Asia/Hong_Kong",   time(9, 30),  time(16, 0)),
    # Singapore
    "SGX":    _Hours("Asia/Singapore",    time(9, 0),   time(17, 0)),
    # India
    "NSE":    _Hours("Asia/Kolkata",      time(9, 15),  time(15, 30)),
    "BSE":    _Hours("Asia/Kolkata",      time(9, 15),  time(15, 30)),
}

_DEFAULT = _MAP["NYSE"]


def is_market_hours(exchange: str = "NYSE") -> bool:
    """Return True if the given exchange is currently within regular trading hours (weekdays only)."""
    spec = _MAP.get(exchange.upper(), _DEFAULT)
    now_local = datetime.now(ZoneInfo(spec.timezone))
    if now_local.weekday() >= 5:  # Saturday=5, Sunday=6
        return False
    t = now_local.time().replace(second=0, microsecond=0)
    return spec.open <= t < spec.close


# Extended pre-market / after-hours windows per exchange.
# Used by is_trading_session() and the manual trade guard.
_EXTENDED: dict[str, tuple[time, time]] = {
    # US equities: 4 AM – 8 PM ET
    "NYSE":   (time(4, 0),  time(20, 0)),
    "NASDAQ": (time(4, 0),  time(20, 0)),
    "ARCA":   (time(4, 0),  time(20, 0)),
    "AMEX":   (time(4, 0),  time(20, 0)),
    "BATS":   (time(4, 0),  time(20, 0)),
    "IEX":    (time(4, 0),  time(20, 0)),
    "SMART":  (time(4, 0),  time(20, 0)),
    "CBOE":   (time(4, 0),  time(20, 0)),
    # Canada: 7 AM – 5 PM ET
    "TSX":    (time(7, 0),  time(17, 0)),
    "TSXV":   (time(7, 0),  time(17, 0)),
    # UK: 7 AM – 5:15 PM London
    "LSE":    (time(7, 0),  time(17, 15)),
    "IOB":    (time(7, 0),  time(17, 15)),
    # Germany: 8 AM – 8 PM Berlin (Xetra extended)
    "XETRA":  (time(8, 0),  time(20, 0)),
    "FWB":    (time(8, 0),  time(20, 0)),
    # France: 7:15 AM – 5:30 PM Paris
    "SBF":    (time(7, 15), time(17, 30)),
    # Netherlands: 7:15 AM – 5:30 PM Amsterdam
    "AEB":    (time(7, 15), time(17, 30)),
    # Australia: pre-open 7 AM, closing auction ends ~4:12 PM; use 7 AM – 5 PM Sydney
    "ASX":    (time(7, 0),  time(17, 0)),
    # Japan: pre-open 8 AM, night session ends 6 AM next day — cap at 4:30 PM Tokyo
    "TSEJ":   (time(8, 0),  time(16, 30)),
    "OSE":    (time(8, 0),  time(16, 30)),
    # Hong Kong: pre-open 9 AM, no significant after-hours; use 9 AM – 5 PM HKT
    "SEHK":   (time(9, 0),  time(17, 0)),
    # Singapore: pre-open 8:30 AM, after-hours 5–7 PM SGT
    "SGX":    (time(8, 30), time(19, 0)),
    # India: pre-market 9 AM, closing session ends ~4 PM IST
    "NSE":    (time(9, 0),  time(16, 0)),
    "BSE":    (time(9, 0),  time(16, 0)),
}


def is_trading_session(exchange: str = "NYSE") -> bool:
    """Return True during pre-market, regular, or after-hours — used by manual trade guards."""
    exch = exchange.upper()
    spec = _MAP.get(exch, _DEFAULT)
    now_local = datetime.now(ZoneInfo(spec.timezone))
    if now_local.weekday() >= 5:
        return False
    t = now_local.time().replace(second=0, microsecond=0)
    # Use extended window if defined, otherwise fall back to regular hours
    ext = _EXTENDED.get(exch)
    if ext:
        return ext[0] <= t < ext[1]
    return spec.open <= t < spec.close
